src/_P084_VEML6070.ino

```diff
@@ -0,0 +1,230 @@
+          String log = F("VEML6070: Not available!");
+          addLog(LOG_LEVEL_INFO, log);
+        }
+        
+        VEML6070_UvTableInit();
+
+        success = true;
+        break;
+      }
+
+    case PLUGIN_READ:
+      {
+        uint16_t uv_raw;
+        double uv_risk, uv_power;
+
+        uv_raw   = VEML6070_ReadUv();             // get UV raw values
+        uv_risk  = VEML6070_UvRiskLevel(uv_raw);  // get UV risk level
+        uv_power = VEML6070_UvPower(uv_risk);     // get UV power in W/m2
+
+        if (isnan(uv_raw) || uv_raw == (-1) || uv_raw == 65535) {
+          String log = F("VEML6070: no data read!");
+          addLog(LOG_LEVEL_INFO, log);
+          UserVar[event->BaseVarIndex + 0] = NAN;
+          UserVar[event->BaseVarIndex + 1] = NAN;
+          UserVar[event->BaseVarIndex + 2] = NAN;
+		      success = false;
+        } else {
+          UserVar[event->BaseVarIndex + 0] = uv_raw;
+          UserVar[event->BaseVarIndex + 1] = uv_risk;
+          UserVar[event->BaseVarIndex + 2] = uv_power;
+          String log = F("VEML6070: UV: ");
+          log += UserVar[event->BaseVarIndex];
+          addLog(LOG_LEVEL_INFO, log);
+		      success = true;
+        }
+
+        success = true;
+        break;
+      }
+
+  }
+  return success;
+}
+
+
+//////////////
+// VEML6070 //
+//////////////
+
+//get UV raw values
+uint16_t VEML6070_ReadUv(void)
+{
+  uint16_t uv_raw = 0;
+  // read high byte
+  if (Wire.requestFrom(VEML6070_ADDR_H, 1) != 1) {
+    return -1;
+  }
+  uv_raw   = Wire.read();
+  uv_raw <<= 8;
+  // read low byte
+  if (Wire.requestFrom(VEML6070_ADDR_L, 1) != 1) {
+    return -1;
+  }
+  uv_raw  |= Wire.read();
+  // high and low done
+  return uv_raw;
+
+}
+
+
+bool VEML6070_Init(byte it)
+{
+  byte error;
+  
+  Wire.begin();
+  Wire.beginTransmission(VEML6070_ADDR_L);
+  Wire.write((it << 2) | 0x02);
+  error = Wire.endTransmission();
+  delay(500);
+
+  return (error == 0);
+}
+
+double VEML6070_UvRiskLevel(uint16_t uv_level)
+{
+  double risk = 0;
+  if (uv_level < uv_risk_map[VEML6070_UV_MAX_INDEX-1]) {
+    risk = (double)uv_level / uv_risk_map[0];
+    // generate uv-risk string
+    if ( (risk >= 0) && (risk <= 2.9) ) { snprintf_P(str_uvrisk_text, sizeof(str_uvrisk_text), D_UV_INDEX_1); }
+    else if ( (risk >= 3.0)  && (risk <= 5.9) )  { snprintf_P(str_uvrisk_text, sizeof(str_uvrisk_text), D_UV_INDEX_2); }
+    else if ( (risk >= 6.0)  && (risk <= 7.9) )  { snprintf_P(str_uvrisk_text, sizeof(str_uvrisk_text), D_UV_INDEX_3); }
+    else if ( (risk >= 8.0)  && (risk <= 10.9) ) { snprintf_P(str_uvrisk_text, sizeof(str_uvrisk_text), D_UV_INDEX_4); }
+    else if ( (risk >= 11.0) && (risk <= 12.9) ) { snprintf_P(str_uvrisk_text, sizeof(str_uvrisk_text), D_UV_INDEX_5); }
+    else if ( (risk >= 13.0) && (risk <= 25.0) ) { snprintf_P(str_uvrisk_text, sizeof(str_uvrisk_text), D_UV_INDEX_6); }
+    else { snprintf_P(str_uvrisk_text, sizeof(str_uvrisk_text), D_UV_INDEX_7); }
+    return risk;
```

Comment:
This is a lot of code to only determine the `D_UV_INDEX_x` value. So please do only that and then a single line to create the text.
Also I don't get it why `uv_risk_map` is needed. You fill it in the init function and it is only read here.
Only the last value of the array is used for comparison and only the first value to compute the risk.
The computations of the values in that array are also not really rocket science, so why waste precious memory on these?
Also the `str_uvrisk_text` array is only set here and never read. So this can be removed or at least has to be done differently.
Now it takes a lot of memory and isn't used.

So at least place this part in comments to keep the ranges. Maybe those can be useful some time.

Apart from that, there is a bug in the code.
For example if the `risk` is 2.91, it will not be assigned an index.
Floating point values must be compared differently.

Reply:
uv_risk_map and calculations and string selection are removed. String was never used